### PR TITLE
Better document mocked-dbus, add missing test methods

### PR DESCRIPTION
**Pull request by cywang117**
~~With this change, dbus is stubbed in isolation and `mock` must be called explicitly to stub dbus methods. The mock-dbus module exposes a set of stubs of methods in `src/lib/dbus.ts` for use during testing.~~

~~Additionally, dbus.getBus instances are now acquired via dependency injection. Previously, on lib/dbus module import, `const bus = dbus.getBus('system')` was called immediately, making isolating the effects of this difficult in the testing environment. The small refactor in lib/dbus functions the same but avoids the import-with-side-effects antipattern.~~

After revisiting the code, the additional overhead for removing imports with side effects from lib/dbus may not be worth it. Instead, I better documented why dbus is stubbed like it is, and added missing methods in the stub (Reboot & PowerOff for example).

This change is needed as part of API test improvements in https://github.com/balena-os/balena-supervisor/pull/1955 (WIP).

Change-type: patch
Signed-off-by: Christina Wang <christina@balena.io>

### CODE REVIEW

File: test/lib/mock-dbus.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import { DBusError, DBusInterface, busType, DBusConnection } from 'dbus';
+import { createSandbox, SinonSandbox, SinonStub } from 'sinon';
+
+import * as dbusLib from '../../src/lib/dbus';
+
+let sandbox: SinonSandbox;
+
+const dbusLibFunctions = Object.keys(dbusLib).filter((k) => k !== 'DbusError');
+
+const stubs: { [key: string]: SinonStub } = {};
```
Comment (pipex) [inline test/lib/mock-dbus.ts:11]: Element in this object are never cleared, meaning data will leak between tests

File: src/lib/dbus.ts [hunk 2]
```diff
@@ -1,37 +1,50 @@
...
 
 import log from './supervisor-console';
 
 export class DbusError extends TypedError {}
 
-const bus = dbus.getBus('system');
-const getInterfaceAsync = Bluebird.promisify(bus.getInterface, {
-	context: bus,
-});
+const getBus = _.memoize((dbusLib: typeof dbus) => dbusLib.getBus('system'));
```
[the remark this answers is not in the document]
Comment (cywang117) [reply src/lib/dbus.ts:10]: The issue with that is that in a test environment, mocha loads all module imports from all spec files before starting tests. When dbus is imported, it gets a system bus immediately (import with side effect) which would cause the test to fail (see test failures in https://github.com/balena-os/balena-supervisor/pull/1961/commits/b775bf3aee7c6b2e6fff9ed6a27532c9b9d4c95b) due to not finding the bus socket in the test env. 

The old mocked-dbus implementation got around this by stubbing `dbus.getBus` on import as well (another import with side effect), but the downside to this was that the mocked dbus module was hard to parse.

If you'd prefer to keep the old implementation in `src/lib/dbus.ts` (i.e. not use memoizee), I can look into this more.

File: test/lib/mock-dbus.ts (hunk 1 above)
Comment (pipex) [reply test/lib/mock-dbus.ts:11]: Oh, I just saw that this object is filled with the list of methods for the stub module

File: src/lib/dbus.ts [hunk 3]
```diff
@@ -1,37 +1,50 @@
...
 async function getSystemdInterface() {
 	try {
-		return await getInterfaceAsync(
+		return await getInterface(
 			'org.freedesktop.systemd1',
 			'/org/freedesktop/systemd1',
 			'org.freedesktop.systemd1.Manager',
 		);
 	} catch (e) {
-		throw new DbusError(e);
```
[the remark this answers is not in the document]
Comment (cywang117) [reply src/lib/dbus.ts:22]: They are of type `DBusError` (note the capital B in DBus), which requires a name and a message on instantiation:
```
export class DBusError extends Error {
    readonly dbusName: string;
    constructor(name: string, message: string);
}
```

I believe the `export class DbusError extends TypedError {}` line in lib/dbus was to avoid this.

File: test/lib/mock-dbus.ts (hunk 1 above)
Comment (cywang117) [reply test/lib/mock-dbus.ts:11]: I added `$DBUS_STUB.reset()` for every stub acquired via `mockDbus.getStub` but that's not the best, you're right. Maybe I'll remove the method stubs entirely here, and manually create dbus stubs in tests that need them

Comment (cywang117) [reply test/lib/mock-dbus.ts:11]: Or maybe `mockDbus.getStub` creates a stub through mockDbus so that the actual dbus does not have to be imported (thus avoiding the import with side effects, if we're not using the memoize implementation)

File: src/lib/dbus.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (cywang117) [reply src/lib/dbus.ts:10]: An example of when src/lib/dbus is imported is in spec 41, the v1 API tests.

[the remark this answers is not in the document]
Comment (pipex) [reply src/lib/dbus.ts:10]: I imagine you do it because you want to mock the interface, but maybe you can try if it is really necessary? Adding the memoization here and in `getInterfaceAsync` will add additional overhead to these calls. While is probably negligible, it still better to make sure we really need it.

File: test/lib/mock-dbus.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			},
+		} as DBusConnection;
+	});
+
+	dbusLibFunctions.forEach((fn) => {
+		stubs[fn] = sandbox.stub(dbusLib, fn as any);
+	});
+};
+
+export const getStub = (method: string) => {
```
Comment (pipex) [inline test/lib/mock-dbus.ts:35]: I understand why we are chosing to stub methods in the dbus library, but it would really be nicer imho if we could actually mock responses coming from the dbus service, avoiding to rely on internal implementation checks.

For instance something like 

```
expect(restartService.callCount).to.equal(2);
```

could be replaced by an integration test that checks that the right call was made to the dbus service. 

This is definitely not the highest priority right now, but is something to keep in mind.

File: src/lib/dbus.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (cywang117) [reply src/lib/dbus.ts]: After looking more into it, the additional overhead added by removing the imports with side effects was not worth the added complexity.. instead, I documented why `dbus.getBus` is stubbed at the root level, and added some missing methods to the stub: https://github.com/balena-os/balena-supervisor/pull/1961/files#diff-eb84a8c34f22bf98ab3a2fff02e87b96d60104060f13bc2954e83ab85d0265d2R5-R17

**Comment by cywang117** [comment]
@balena-ci I self-certify!
